src/nat/middleware/red_teaming/red_teaming_middleware.py

```
@@ -0,0 +1,342 @@
+    def _apply_payload_to_function_value(self, value: Any) -> Any:
+        if self._call_limit is not None and self._call_count >= self._call_limit:
+            logger.warning("Call limit reached for red teaming middleware. "
+                           "Not applying attack payload to value: %s",
+                           value)
+            return value
+        if isinstance(value, list | dict | BaseModel):
+            modified_value = self._apply_payload_to_complex_type(value)
+        elif isinstance(value, str | int | float):
+            modified_value = self._apply_payload_to_simple_type(value, self._attack_payload, self._payload_placement)
+        else:
+            raise ValueError(f"Unsupported function input/output type: {type(value).__name__}")
```

Ruff (0.14.8)

Prefer `TypeError` exception for invalid type

(TRY004)